salt/utils/pycrypto.py

```diff
@@ -103,31 +103,36 @@ def gen_hash(crypt_salt=None, password=None, algorithm=None, force=False):
     """
     Generate /etc/shadow hash
     """
+    log.debug("gen_hash start with: %s", (crypt_salt, password, algorithm, force))
     if password is None:
         password = secure_password()
 
+    if algorithm is None:
+        # use the most secure natively supported method
+        algorithm = crypt.methods[0].name.lower() if HAS_CRYPT else known_methods[0]
```

Comment:
We should either:

1.  keep the `known_methods[0]` here, and delete `if algorithm is None` in `_fallback_gen_hash`

or

2. `if algorithm is None and HAS_CRYPT`, and modify `if algorithm is None` in `_fallback_gen_hash` to set algorithm to `known_methods[0]`.

Reply:
I'll re-check, thank you.